util/timeutil/time.go

```diff
@@ -0,0 +1,134 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package timeutil
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// init initializes `locCache`.
+func init() {
+	LocCache = &locCache{}
+	LocCache.locMap = make(map[string]*time.Location)
+}
+
+// LocCache is a simple cache policy to improve the performance of 'time.LoadLocation'.
+var LocCache *locCache
+var localStr string
+var localOnce sync.Once
+var zoneSources = []string{
+	"/usr/share/zoneinfo/",
+	"/usr/share/lib/zoneinfo/",
+	"/usr/lib/locale/TZ/",
+}
+
+// locCache is a simple map with lock. It stores all used timezone during the lifetime of tidb instance.
+// Talked with Golang team about whether they can have some forms of cache policy available for programmer,
+// they suggests that only programmers knows which one is best for their use case.
+// For detail, please refer to: https://github.com/golang/go/issues/26106
+type locCache struct {
+	sync.RWMutex
+	// locMap stores locations used in past and can be retrieved by a timezone's name.
+	locMap map[string]*time.Location
+}
+
+func initLocalStr() {
+	// consult $TZ to find the time zone to use.
+	// no $TZ means use the system default /etc/localtime.
+	// $TZ="" means use UTC.
+	// $TZ="foo" means use /usr/share/zoneinfo/foo.
+	tz, ok := syscall.Getenv("TZ")
+	if ok && tz != "" {
+		for _, source := range zoneSources {
+			if _, err := os.Stat(source + tz); os.IsExist(err) {
+				localStr = tz
+				break
+			}
+		}
+		// } else if tz == "" {
+		// localStr = "UTC"
+	} else {
+		path, err := filepath.EvalSymlinks("/etc/localtime")
+		if err == nil {
+			localStr, err = getTZNameFromFileName(path)
+			if err == nil {
+				return
+			}
+			log.Errorln(err)
+		}
+		log.Errorln(err)
+	}
+	localStr = "System"
+}
+
+// getTZNameFromFileName gets IANA timezone name from zoneinfo path.
+// TODO It will be refined later. This is just a quick fix.
+func getTZNameFromFileName(path string) (string, error) {
+	// phase1 only support read /etc/localtime which is a softlink to zoneinfo file
+	substr := "zoneinfo"
+	if strings.Contains(path, substr) {
+		idx := strings.Index(path, substr)
+		return string(path[idx+len(substr)+1:]), nil
+	}
+	return "", errors.New(fmt.Sprintf("path %s is not supported", path))
+}
+
+// Local returns time.Local's IANA timezone name.
+func Local() *time.Location {
+	localOnce.Do(initLocalStr)
+	loc, err := LoadLocation(localStr)
+	if err != nil {
+		return time.Local
+	}
+	return loc
+}
+
+// getLoc first trying to load location from a cache map. If nothing found in such map, then call
+// `time.LocadLocation` to get a timezone location. After trying both way, an error wil be returned
+//  if valid Location is not found.
+func (lm *locCache) getLoc(name string) (*time.Location, error) {
+	if name == "System" {
```

Comment:
I think the default TZ in mysql is "System", just returning Local may defeat the purpose of this PR, which expects to replace Local with the actual timezone name.

Reply:
Yes. `scCtx.GetSessionVars.Location` calls `timeutil.Local` which is trying their best to get system's timezone. When `system` is passed into this function, it means we failed to get timezone from `TZ` or `/etc/localtime`.